homeassistant/components/powerwall/__init__.py

```diff
@@ -0,0 +1,159 @@
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=conf,
+        )
+    )
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Tesla Powerwall from a config entry."""
+
+    entry_id = entry.entry_id
+
+    hass.data[DOMAIN].setdefault(entry_id, {})
+    power_wall = PowerWall(entry.data[CONF_IP_ADDRESS])
+
+    def _call_site_info(power_wall):
+        """Wrap site_info to be a callable."""
+        return power_wall.site_info
+
+    site_info = None
+    try:
+        site_info = await hass.async_add_executor_job(_call_site_info, power_wall)
+    except (PowerWallUnreachableError, ApiError, ConnectionError):
+        raise ConfigEntryNotReady
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        return await hass.async_add_executor_job(_fetch_powerwall_data, power_wall)
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="Powerwall site",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=UPDATE_INTERVAL),
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        POWERWALL_OBJECT: power_wall,
+        POWERWALL_COORDINATOR: coordinator,
+        POWERWALL_SITE_INFO: site_info,
+        POWERWALL_IP_ADDRESS: entry.data[CONF_IP_ADDRESS],
+    }
+
+    await coordinator.async_refresh()
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+def _fetch_powerwall_data(power_wall):
+    """Process and update powerwall data."""
+    meters = power_wall.meters
+    return {
+        POWERWALL_API_CHARGE: power_wall.charge,
+        POWERWALL_API_SITEMASTER: power_wall.sitemaster,
+        POWERWALL_API_METERS: {
+            meter: MetersResponse(meters[meter]) for meter in meters
+        },
+        POWERWALL_API_GRID_STATUS: power_wall.grid_status,
+    }
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class PowerWallEntity(Entity):
+    """Base class for powerwall entities."""
+
+    def __init__(self, coordinator, site_info, ip_address):
+        """Initialize the sensor."""
+        super().__init__()
+        self._coordinator = coordinator
+        self._site_info = site_info
+        self._site_name = site_info[POWERWALL_SITE_NAME]
+        self._ip_address = ip_address
+
+    @property
+    def device_info(self):
+        """Powerwall device info."""
+        return {
+            "identifiers": {(DOMAIN, self._ip_address)},
```

Comment:
IP address is not stable so we can't use it as id.

Reply:
Changed